backend/hasher/hasher.go

```diff
@@ -0,0 +1,550 @@
+//go:build !plan9 && !js
+// +build !plan9,!js
+
+// Package hasher implements a checksum handling overlay backend
+package hasher
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/cache"
+	"github.com/rclone/rclone/fs/config/configmap"
+	"github.com/rclone/rclone/fs/config/configstruct"
+	"github.com/rclone/rclone/fs/fspath"
+	"github.com/rclone/rclone/fs/hash"
+	"go.etcd.io/bbolt"
+)
+
+// Register with Fs
+func init() {
+	fs.Register(&fs.RegInfo{
+		Name:        "hasher",
+		Description: "Hasher will cache hashes",
+		NewFs:       NewFs,
+		CommandHelp: commandHelp,
+		Options: []fs.Option{{
+			Name:     "remote",
+			Required: true,
+			Help:     "Remote to cache hashes for, like myremote:mypath",
+		}, {
+			Name:     "hash_names",
+			Default:  "md5",
+			Advanced: false,
+			Help:     "Space/comma separated list of checksum types to support.",
+		}, {
+			Name:     "auto_size",
+			Advanced: false,
+			Default:  fs.SizeSuffix(1024), // 1K
+			Help:     "Auto-update checksum for files smaller than this size.",
+		}, {
+			Name:     "max_age",
+			Advanced: false,
+			Default:  fs.DurationOff,
+			Help:     "Maximum time to keep checksums in cache. 0 = no-cache, off = forever",
+		}, {
+			Name:     "lock_time",
+			Advanced: true,
+			Default:  fs.Duration(time.Second),
+			Help:     "Maximum time interval to keep cache database locked by process",
```

Comment:
I see - this is to help with multi-process access. I wonder how well it works in practice?

Reply:
During tests I ran 2-3 rclone instances doing `rclone sha1sum /some/local/dir123` with `-vv`, touched random files under test dir  and watched them reopen bolt periodically due to changed fingerprint and following cache updates.

As a stress test for db locking I ran dump (I decided that this maintenance operation can lock db as a whole) in parallel with above operations and watched low-level retries.

I still wonder how I can convert it into backend unit tests.

In practice I usually have at most 1 concurrent rclone instance doing checksum ops and multiple rclone mounts that don't need checksums.